Source/Core/Core/PowerPC/Expression.cpp

```diff
@@ -0,0 +1,229 @@
+{
+  return PowerPC::HostRead_U8(address);
+}
+
+template <>
+u16 HostRead(u32 address)
+{
+  return PowerPC::HostRead_U16(address);
+}
+
+template <>
+u32 HostRead(u32 address)
+{
+  return PowerPC::HostRead_U32(address);
+}
+
+template <>
+u64 HostRead(u32 address)
+{
+  return PowerPC::HostRead_U64(address);
+}
+
+template <>
+void HostWrite(u8 var, u32 address)
+{
+  PowerPC::HostWrite_U8(var, address);
+}
+
+template <>
+void HostWrite(u16 var, u32 address)
+{
+  PowerPC::HostWrite_U16(var, address);
+}
+
+template <>
+void HostWrite(u32 var, u32 address)
+{
+  PowerPC::HostWrite_U32(var, address);
+}
+
+template <>
+void HostWrite(u64 var, u32 address)
+{
+  PowerPC::HostWrite_U64(var, address);
+}
+
+template <typename T, typename U = T>
+static double HostReadFunc(expr_func* f, vec_expr_t* args, void* c)
+{
+  if (vec_len(args) != 1)
+    return 0;
+  u32 address = static_cast<u32>(expr_eval(&vec_nth(args, 0)));
+  return Common::BitCast<T>(HostRead<U>(address));
+}
+
+template <typename T, typename U = T>
+static double HostWriteFunc(expr_func* f, vec_expr_t* args, void* c)
+{
+  if (vec_len(args) != 2)
+    return 0;
+  T var = static_cast<T>(expr_eval(&vec_nth(args, 0)));
+  u32 address = static_cast<u32>(expr_eval(&vec_nth(args, 1)));
+  HostWrite<U>(Common::BitCast<U>(var), address);
+  return var;
+}
+
+template <typename T, typename U = T>
+static double CastFunc(expr_func* f, vec_expr_t* args, void* c)
+{
+  if (vec_len(args) != 1)
+    return 0;
+  return Common::BitCast<T>(static_cast<U>(expr_eval(&vec_nth(args, 0))));
+}
+
+static expr_func g_expr_funcs[] = {{"read_u8", HostReadFunc<u8>},
+                                   {"read_s8", HostReadFunc<s8, u8>},
+                                   {"read_u16", HostReadFunc<u16>},
+                                   {"read_s16", HostReadFunc<s16, u16>},
+                                   {"read_u32", HostReadFunc<u32>},
+                                   {"read_s32", HostReadFunc<s32, u32>},
+                                   {"read_f32", HostReadFunc<float, u32>},
+                                   {"read_f64", HostReadFunc<double, u64>},
+                                   {"write_u8", HostWriteFunc<u8>},
+                                   {"write_u16", HostWriteFunc<u16>},
+                                   {"write_u32", HostWriteFunc<u32>},
+                                   {"write_f32", HostWriteFunc<float, u32>},
+                                   {"write_f64", HostWriteFunc<double, u64>},
+                                   {"u8", CastFunc<u8>},
+                                   {"s8", CastFunc<s8, u8>},
+                                   {"u16", CastFunc<u16>},
+                                   {"s16", CastFunc<s16, u16>},
+                                   {"u32", CastFunc<u32>},
+                                   {"s32", CastFunc<s32, u32>},
+                                   {}};
+
+const std::unordered_map<std::string_view, u32&> g_register_vars = {
```

Comment:
Please consider using a std::array here instead of unordered map to avoid static allocating constructors. This is a small enough data set that linear lookup is more or less fine.

Ditto with g_float_register_vars

Reply:
Would population of the map on first use be an acceptable alternative? Is your principal concern app startup time?

Reply:
No.

We generally try to avoid using allocating structures at file scope when they're avoidable. Given this is a small set of entries (< 100) it is fine to use a std::array here and make use of `std::find_if`

Reply:
They're gone. Variable binding now occurs at expression compile time, so no string comparisons are necessary at evaluation time.

Reply:
I opted to remove them entirely. Variable binding is now done at expression compile time (as opposed to evaluation time) so fast lookup is not such a concern.